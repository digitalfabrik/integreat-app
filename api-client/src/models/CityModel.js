// @flow

class CityModel {
  _name: string
  _code: string
  _live: boolean
  _eventsEnabled: boolean
  _offersEnabled: boolean
  _pushNotificationsEnabled: boolean
  _tunewsEnabled: boolean
  _sortingName: string
  _prefix: ?string
  _latitude: number | null
  _longitude: number | null
  _aliases: { [alias: string]: {| longitude: number, latitude: number |} } | null

  constructor (params: {|
    name: string, code: string, live: boolean, eventsEnabled: boolean,
    offersEnabled: boolean, pushNotificationsEnabled: boolean, tunewsEnabled: boolean,
    sortingName: string, prefix: ?string, latitude: number | null, longitude: number | null,
    aliases: { [alias: string]: {| longitude: number, latitude: number |} } | null
  |}) {
    this._name = params.name
    this._code = params.code
    this._live = params.live
    this._eventsEnabled = params.eventsEnabled
    this._offersEnabled = params.offersEnabled
    this._pushNotificationsEnabled = params.pushNotificationsEnabled
    this._tunewsEnabled = params.tunewsEnabled
    this._sortingName = params.sortingName
    this._prefix = params.prefix
    this._latitude = params.latitude
    this._longitude = params.longitude
    this._aliases = params.aliases
  }

  get live (): boolean {
    return this._live
  }

  get name (): string {
    return this._name
  }

  get code (): string {
    return this._code
  }

  get sortingName (): string {
    return this._sortingName
  }

  get eventsEnabled (): boolean {
    return this._eventsEnabled
  }

  get offersEnabled (): boolean {
    return this._offersEnabled
  }

  get pushNotificationsEnabled (): boolean {
    return this._pushNotificationsEnabled
  }

  get tunewsEnabled (): boolean {
    return this._tunewsEnabled
  }

  get sortCategory (): string {
    return this._sortingName.charAt(0)
  }

  get prefix (): ?string {
    return this._prefix
  }

  get longitude (): number | null {
    return this._longitude
  }

  get latitude (): number | null {
    return this._latitude
  }

  get aliases (): { [alias: string]: {| longitude: number, latitude: number |} } | null {
    return this._aliases
  }

  static findCityName (cities: Array<CityModel>, code: string): string {
    const city = cities.find(city => city.code === code)
    return city ? city.name : code
  }
}

export default CityModel
