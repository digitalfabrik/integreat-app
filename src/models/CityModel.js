// @flow

class CityModel {
  _name: string
  _code: string
  _live: boolean
  _eventsEnabled: boolean
  _extrasEnabled: boolean
  _sortingName: string
  _prefix: ?string
  _latitude: number | null
  _longitude: number | null
  _aliases: { [alias: string]: {| longitude: number, latitude: number |} } | null

  constructor (params: {|
    name: string, code: string, live: boolean, eventsEnabled: boolean, extrasEnabled: boolean,
    sortingName: string, prefix: ?string, latitude: number | null, longitude: number | null,
    aliases: { [alias: string]: {| longitude: number, latitude: number |} } | null
  |}) {
    this._name = params.name
    this._code = params.code
    this._live = params.live
    this._eventsEnabled = params.eventsEnabled
    this._extrasEnabled = params.extrasEnabled
    this._sortingName = params.sortingName
    this._prefix = params.prefix
    this._latitude = params.latitude
    this._longitude = params.longitude
    this._aliases = params.aliases
  }

  get live (): boolean {
    return this._live
  }

  get name (): string {
    return this._name
  }

  get code (): string {
    return this._code
  }

  get sortingName (): string {
    return this._sortingName
  }

  get eventsEnabled (): boolean {
    return this._eventsEnabled
  }

  get extrasEnabled (): boolean {
    return this._extrasEnabled
  }

  get sortCategory (): string {
    return this._sortingName.charAt(0)
  }

  get prefix (): ?string {
    return this._prefix
  }

  get longitude (): number | null {
    return this._longitude
  }

  get latitude (): number | null {
    return this._latitude
  }

  get aliases (): { [alias: string]: {| longitude: number, latitude: number |} } | null {
    return this._aliases
  }

  static findCityName (cities: Array<CityModel>, code: string): string {
    const city = cities.find(city => city.code === code)
    return city ? city.name : code
  }
}

export default CityModel
