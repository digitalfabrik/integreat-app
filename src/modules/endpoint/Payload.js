// @flow

import isUrl from 'is-url'

/**
 * The payload gets stored in the redux store and holds the information about a fetch
 */
class Payload<T> {
  _isFetching: boolean
  _data: ?T
  _error: ?Error
  _requestUrl: ?string
  _fetchDate: number

  constructor (
    isFetching: boolean,
    requestUrl: ?string = null,
    data: ?T = null,
    error: ?Error = null,
    fetchDate: number = Date.now()
  ) {
    this._isFetching = isFetching
    this._fetchDate = fetchDate
    this._error = error
    this._requestUrl = requestUrl
    this._data = data

    if (requestUrl) {
      if (!isUrl(requestUrl)) {
        throw new Error('requestUrl must be a valid URL')
      }
    }

    if (error && data) {
      throw new Error('data and error can not be set at the same time')
    }
  }

  get fetchDate (): number {
    return this._fetchDate
  }

  get isFetching (): boolean {
    return this._isFetching
  }

  get data (): ?T {
    return this._data
  }

  get error (): ?Error {
    return this._error
  }

  get requestUrl (): ?string {
    return this._requestUrl
  }
}

export default Payload
