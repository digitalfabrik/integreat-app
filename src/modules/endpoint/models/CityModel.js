// @flow

class CityModel {
  _name: string
  _code: string
  _live: boolean
  _eventsEnabled: boolean
  _extrasEnabled: boolean
  _sortingName: string

  constructor (params: {|name: string, code: string, live: boolean, eventsEnabled: boolean, extrasEnabled: boolean,
    sortingName: string|}) {
    this._name = params.name
    this._code = params.code
    this._live = params.live
    this._eventsEnabled = params.eventsEnabled
    this._extrasEnabled = params.extrasEnabled
    this._sortingName = params.sortingName
  }

  get live (): boolean {
    return this._live
  }

  get name (): string {
    return this._name
  }

  get code (): string {
    return this._code
  }

  get sortingName (): string {
    return this._sortingName
  }

  get eventsEnabled (): boolean {
    return this._eventsEnabled
  }

  get extrasEnabled (): boolean {
    return this._extrasEnabled
  }

  get sortCategory (): string {
    return this._sortingName.charAt(0)
  }

  static findCityName (cities: Array<CityModel>, code: string): string {
    const city = cities.find(city => city.code === code)
    return city ? city.name : code
  }
}

export default CityModel
