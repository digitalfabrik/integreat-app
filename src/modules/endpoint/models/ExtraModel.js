export default class ExtraModel {
  constructor ({alias, name, path, thumbnail}) {
    this._alias = alias
    this._name = name
    this._path = path
    this._thumbnail = thumbnail
  }

  get alias () {
    return this._alias
  }

  get thumbnail () {
    return this._thumbnail
  }

  get name () {
    return this._name
  }

  get path () {
    return this._path
  }
}
