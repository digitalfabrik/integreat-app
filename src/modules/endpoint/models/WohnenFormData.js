// @flow

import type Moment from 'moment'

export interface LandlordType {
  +firstName: string,
  +lastName: string,
  +phone: string
}

export interface AccommodationType {
  +ofRooms: Array<string>,
  +title: string,
  +location: string,
  +totalArea: number,
  +totalRooms: number,
  +moveInDate: Moment,
  +ofRoomsDiff: Array<string>
}

export interface CostsType {
  +ofRunningServices: Array<string>,
  +ofAdditionalServices: Array<string>,
  +baseRent: number,
  +runningCosts: number,
  +hotWaterInHeatingCosts: boolean,
  +additionalCosts: number,
  +ofRunningServicesDiff: Array<string>,
  +ofAdditionalServicesDiff: Array<string>
}

export default class WohnenFormData {
  _landlord: LandlordType
  _accommodation: AccommodationType
  _costs: CostsType

  constructor (landlord: LandlordType, accommodation: AccommodationType, costs: CostsType) {
    this._landlord = landlord
    this._accommodation = accommodation
    this._costs = costs
  }

  get landlord (): LandlordType {
    return this._landlord
  }

  get accommodation (): AccommodationType {
    return this._accommodation
  }

  get costs (): CostsType {
    return this._costs
  }
}
