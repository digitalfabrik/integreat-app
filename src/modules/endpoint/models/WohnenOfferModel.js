// @flow

import type Moment from 'moment'

class WohnenOfferModel {
  _email: string
  _createdDate: Moment
  _formData: *

  constructor (params: {
    email: string,
    createdDate: Moment,
    formData: *
  }) {
    this._email = params.email
    this._createdDate = params.createdDate
    this._formData = params.formData
  }

  get email (): string {
    return this._email
  }

  get createdDate (): Moment {
    return this._createdDate
  }

  get formData (): * {
    return this._formData
  }
}

export default WohnenOfferModel
