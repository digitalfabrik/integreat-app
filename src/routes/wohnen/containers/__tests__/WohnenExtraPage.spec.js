// @flow

import { shallow } from 'enzyme'
import React from 'react'

import ExtraModel from 'modules/endpoint/models/ExtraModel'
import ConnectedWohnenExtraPage, { WohnenExtraPage } from '../WohnenExtraPage'
import configureMockStore from 'redux-mock-store'
import CityModel from 'modules/endpoint/models/CityModel'
import WohnenOfferModel from 'modules/endpoint/models/WohnenOfferModel'
import moment from 'moment'
import WohnenFormData from 'modules/endpoint/models/WohnenFormData'
import Hashids from 'hashids'
import Payload from 'modules/endpoint/Payload'

describe('WohnenExtraPage', () => {
  const city = 'augsburg'
  const language = 'de'

  const wohnenExtra = new ExtraModel({
    alias: 'wohnen', path: 'path to fetch offers from', title: 'Raumfrei', thumbnail: 'xy'
  })

  const extras = [wohnenExtra]

  const cities = [
    new CityModel({
      name: 'Augsburg',
      code: 'augsburg',
      live: true,
      eventsEnabled: true,
      extrasEnabled: false,
      sortingName: 'Augsburg'
    })
  ]

  const offer = new WohnenOfferModel({
    email: 'mail@mail.com',
    createdDate: moment('2018-07-24T00:00:00.000Z'),
    formDataType: WohnenFormData,
    formData: new WohnenFormData(
      {
        firstName: 'Max',
        lastName: 'Ammann',
        phone: ''
      },
      {
        ofRooms: ['kitchen', 'child2', 'child1', 'bed'],
        title: 'Test Angebot',
        location: 'Augsburg',
        totalArea: 120,
        totalRooms: 4,
        moveInDate: moment('2018-07-19T15:35:12.000Z'),
        ofRoomsDiff: ['bath', 'wc', 'child3', 'livingroom', 'hallway', 'store', 'basement', 'balcony']
      },
      {
        ofRunningServices: ['chimney', 'other'],
        ofAdditionalServices: ['garage'],
        baseRent: 1000,
        runningCosts: 1200,
        hotWaterInHeatingCosts: true,
        additionalCosts: 200,
        ofRunningServicesDiff: ['heating', 'water', 'garbage'],
        ofAdditionalServicesDiff: []
      })
  })
  const offerHash = new Hashids().encode(offer.email.length, offer.createdDate.milliseconds())

  const offers = [offer]

  it('should render list if no hash is supplied', () => {
    const wohnenPage = shallow(
      <WohnenExtraPage offers={offers}
                       city={city}
                       language={language}
                       extras={[wohnenExtra]}
                       cities={cities} />
    )
    expect(wohnenPage).toMatchSnapshot()
  })

  it('should render detailed offer if hash is supplied', () => {
    const wohnenPage = shallow(
      <WohnenExtraPage offers={offers}
                       city={city}
                       language={language}
                       offerHash={offerHash}
                       extras={[wohnenExtra]}
                       cities={cities} />
    )
    expect(wohnenPage).toMatchSnapshot()
  })

  it('should render failure offer if offer is not found', () => {
    const extrasPage = shallow(
      <WohnenExtraPage offers={offers}
                       city={city}
                       language={language}
                       offerHash={'invalid hash'}
                       extras={[wohnenExtra]}
                       cities={cities} />
    )
    expect(extrasPage).toMatchSnapshot()
  })

  it('should render failure city does not support offer', () => {
    const wohnenPage = shallow(
      <WohnenExtraPage offers={offers}
                       city={city}
                       language={language}
                       offerHash={offerHash}
                       extras={[]}
                       cities={cities} />
    )
    expect(wohnenPage).toMatchSnapshot()
  })

  it('should render spinner if offers are not ready', () => {
    const wohnenPage = shallow(
      <WohnenExtraPage offers={null}
                       city={city}
                       language={language}
                       offerHash={offerHash}
                       extras={[wohnenExtra]}
                       cities={cities} />
    )
    expect(wohnenPage).toMatchSnapshot()
  })

  it('should map state to props', () => {
    const offerHash = 'hASH'
    const location = {payload: {language, city, offerHash}}

    const mockStore = configureMockStore()
    const store = mockStore({
      location: location,
      extras: new Payload(false, null, extras),
      wohnen: new Payload(false, null, offers),
      cities: new Payload(false, null, cities)
    })

    const wohnenPage = shallow(
      <ConnectedWohnenExtraPage store={store} />
    )

    expect(wohnenPage.props()).toMatchObject({
      language,
      city,
      offerHash,
      extras,
      cities,
      offers
    })
  })
})
