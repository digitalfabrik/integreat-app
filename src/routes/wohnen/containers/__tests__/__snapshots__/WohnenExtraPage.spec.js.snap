// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WohnenExtraPage should render detailed offer if hash is supplied 1`] = `
<Translate(OfferDetail)
  offer={
    WohnenOfferModel {
      "_createdDate": "2018-07-24T00:00:00.000Z",
      "_email": "mail@mail.com",
      "_formData": WohnenFormData {
        "_accommodation": Object {
          "location": "Augsburg",
          "moveInDate": "2018-07-19T15:35:12.000Z",
          "ofRooms": Array [
            "kitchen",
            "child2",
            "child1",
            "bed",
          ],
          "ofRoomsDiff": Array [
            "bath",
            "wc",
            "child3",
            "livingroom",
            "hallway",
            "store",
            "basement",
            "balcony",
          ],
          "title": "Test Angebot",
          "totalArea": 120,
          "totalRooms": 4,
        },
        "_costs": Object {
          "additionalCosts": 200,
          "baseRent": 1000,
          "hotWaterInHeatingCosts": true,
          "ofAdditionalServices": Array [
            "garage",
          ],
          "ofAdditionalServicesDiff": Array [],
          "ofRunningServices": Array [
            "chimney",
            "other",
          ],
          "ofRunningServicesDiff": Array [
            "heating",
            "water",
            "garbage",
          ],
          "runningCosts": 1200,
        },
        "_landlord": Object {
          "firstName": "Max",
          "lastName": "Ammann",
          "phone": "",
        },
      },
    }
  }
/>
`;

exports[`WohnenExtraPage should render failure city does not support offer 1`] = `
<FailureSwitcher
  error={[Error: The Wohnen extra is not supported.]}
/>
`;

exports[`WohnenExtraPage should render failure offer if offer is not found 1`] = `
<FailureSwitcher
  error={[Error: Angebot nicht gefunden.]}
/>
`;

exports[`WohnenExtraPage should render list if no hash is supplied 1`] = `
<Fragment>
  <Caption
    title="Raumfrei"
  />
  <List
    items={
      Array [
        WohnenOfferModel {
          "_createdDate": "2018-07-24T00:00:00.000Z",
          "_email": "mail@mail.com",
          "_formData": WohnenFormData {
            "_accommodation": Object {
              "location": "Augsburg",
              "moveInDate": "2018-07-19T15:35:12.000Z",
              "ofRooms": Array [
                "kitchen",
                "child2",
                "child1",
                "bed",
              ],
              "ofRoomsDiff": Array [
                "bath",
                "wc",
                "child3",
                "livingroom",
                "hallway",
                "store",
                "basement",
                "balcony",
              ],
              "title": "Test Angebot",
              "totalArea": 120,
              "totalRooms": 4,
            },
            "_costs": Object {
              "additionalCosts": 200,
              "baseRent": 1000,
              "hotWaterInHeatingCosts": true,
              "ofAdditionalServices": Array [
                "garage",
              ],
              "ofAdditionalServicesDiff": Array [],
              "ofRunningServices": Array [
                "chimney",
                "other",
              ],
              "ofRunningServicesDiff": Array [
                "heating",
                "water",
                "garbage",
              ],
              "runningCosts": 1200,
            },
            "_landlord": Object {
              "firstName": "Max",
              "lastName": "Ammann",
              "phone": "",
            },
          },
        },
      ]
    }
    noItemsMessage="noOffersAvailable"
    renderItem={[Function]}
  />
</Fragment>
`;
